text_extensions_for_pandas/io/tokenization.py

```diff
@@ -70,28 +74,171 @@ def make_bert_tokens(target_text: str, tokenizer) -> pd.DataFrame:
+    begins = offset_df["begin"].mask(special_tokens_mask, other=ends).astype("int32")
 
     # Create char and token span arrays
     char_spans = CharSpanArray(target_text, begins, ends)
-    token_spans = TokenSpanArray(char_spans,
-                                 np.arange(len(char_spans)),
-                                 np.arange(1, len(char_spans) + 1))
-
-    token_features = pd.DataFrame({
-        "id": special_tokens_mask.index,
-        # Use values instead of series because different indexes
-        "char_span": pd.Series(char_spans).values,
-        "token_span": pd.Series(token_spans).values,
-        "input_id": tokenized_result["input_ids"],
-        "token_type_id": tokenized_result["token_type_ids"],
-        "attention_mask": tokenized_result["attention_mask"],
-        "special_tokens_mask": special_tokens_mask,
-    })
+    token_spans = TokenSpanArray(
+        char_spans, np.arange(len(char_spans)), np.arange(1, len(char_spans) + 1)
+    )
+
+    token_features = pd.DataFrame(
+        {
+            "id": special_tokens_mask.index,
+            # Use values instead of series because different indexes
+            "char_span": pd.Series(char_spans).values,
+            "token_span": pd.Series(token_spans).values,
+            "input_id": tokenized_result["input_ids"],
+            "token_type_id": tokenized_result["token_type_ids"],
+            "attention_mask": tokenized_result["attention_mask"],
+            "special_tokens_mask": special_tokens_mask,
+        }
+    )
 
     return token_features
+
+
+def seq_to_windows(
+    seq: np.ndarray, overlap: int, non_overlap: int
+) -> Dict[str, np.ndarray]:
+    """
+    Convert a variable-length sequence into a set of fixed length windows,
+    adding padding as necessary.
+
+    Usually this function is used to prepare batches of BERT tokens to feed
+    to a BERT model.
+
+    :param seq: Original variable length sequence, as a 1D numpy array
+    :param overlap: How much overlap there should be between adjacent windows
+    :param non_overlap: How much non-overlapping content between the overlapping
+     regions there should be at the middle of each window?
+
+    :returns: A dictionary with two entries:
+      * `input_ids`: 2D `np.ndarray` of fixed-length windows
+      * `attention_masks`: 2D `np.ndarray` of attention masks (1 for
+        tokens that are NOT masked, 0 for tokens that are masked)
+        to feed into your favorite BERT-like embedding generator.
+    """
+    if len(seq.shape) != 1:
+        raise ValueError(f"Input array must be 1D; got shape {seq.shape}")
+    window_length, pre_padding, post_padding = _compute_padding(
+        len(seq), overlap, non_overlap
+    )
+
+    # First generate the windows as a padded flat arrays.
+    padded_length = len(seq) + pre_padding + post_padding
+    buf = np.zeros(shape=[padded_length], dtype=seq.dtype)
+    buf[pre_padding : pre_padding + len(seq)] = seq
```

Comment:
would `np.concatenate` be better than making a copy of `seq`?

Reply:
I don't think there would be any performance benefit to using `np.concatenate` here. As far as I'm aware, `np.concatenate` always allocates a fresh, contiguous buffer and copies the arrays being concatenated into that buffer; see source code at https://github.com/numpy/numpy/blob/50ce0fce70cac779919d97578381a9d762f42594/numpy/core/src/multiarray/multiarraymodule.c#L363